effekt/shared/src/main/scala/effekt/core/StaticArguments.scala

```diff
@@ -0,0 +1,242 @@
+
+    val calleeType = BlockType.Function(
+      blockLit.tparams,
+      dropStatic(staticC, blockLit.cparams),
+      dropStatic(staticV, blockLit.vparams).map(_.tpe),
+      dropStatic(staticB, blockLit.bparams).map(_.tpe),
+      blockLit.tpe match
+        case BlockType.Function(_, _, _, _, result) => result
+        case _ => ??? // impossible!
+    )
+
+    // TODO: Where do we get the capture set?
+    val worker: Block.BlockVar = BlockVar(Id(definition.id.name.name + "_worker"), calleeType, Set())
+    ctx.workers(definition.id) = worker
+
+    // push to the stack to detect recursion, since recursive calls will need to call the worker
+    def rewriteBody(body: Stmt) =
+      val before = ctx.stack
+      ctx.stack = definition.id :: ctx.stack
+      val newBody = rewrite(body)
+      ctx.stack = before
+      newBody
+    
+    // fresh params for the wrapper function and its invocation
+    // note: only freshen params if not static to prevent duplicates
+    val freshCparams = staticC.map { static =>
+      blockLit.cparams.zip(static).map((c, bool) => if (bool) c else Id(c.name.name))
+    }.getOrElse(blockLit.cparams)
+    val freshVparams = staticV.map { static =>
+      blockLit.vparams.zip(static).map((v, bool) => ValueParam(if (bool) v.id else Id(v.id.name.name), v.tpe)).asInstanceOf[List[Param.ValueParam]]
+    }.getOrElse(blockLit.vparams)
+    val freshBparams = staticB.map { static =>
+      blockLit.bparams.zip(static).map((b, bool) => BlockParam(if (bool) b.id else Id(b.id.name.name), b.tpe, b.capt)).asInstanceOf[List[Param.BlockParam]]
+    }.getOrElse(blockLit.bparams)
+
+    Definition.Def(definition.id, BlockLit(
+      blockLit.tparams, // TODO: Do we need to freshen these as well?
+      freshCparams,
+      freshVparams,
+      freshBparams,
+      Scope(List(Definition.Def(worker.id, BlockLit(
+        blockLit.tparams,
+        dropStatic(staticC, blockLit.cparams),
+        dropStatic(staticV, blockLit.vparams),
+        dropStatic(staticB, blockLit.bparams),
+        rewriteBody(blockLit.body)
+      ))), App(
+        worker,
+        // TODO: These conversions to types are a bit hacky, are there better ways?
+        blockLit.tparams.map(t => ValueType.Var(t)),
+        dropStatic(staticV, freshVparams.map(v => ValueVar(v.id, v.tpe))),
+        dropStatic(staticB, freshBparams.map(b => BlockVar(b.id, b.tpe, b.capt)))
+      ))
+    ))
+  
+  def rewrite(definitions: List[Definition])(using ctx: StaticArgumentsContext): List[Definition] =
+    definitions.collect {
+      case d @ Definition.Def(id, block) if ctx.staticC.contains(id) || ctx.staticV.contains(id) || ctx.staticB.contains(id)
+        => wrapDefinition(d)
+      case Definition.Def(id, block) => Definition.Def(id, rewrite(block))
+      case Definition.Let(id, tpe, binding) => Definition.Let(id, tpe, rewrite(binding))
+    }
+
+  def rewrite(d: Definition)(using StaticArgumentsContext): Definition = d match {
```

Comment:
Same question here. Can you use [`Rewrite`](https://github.com/effekt-lang/effekt/blob/042dc2d2691e54c76502ec104a50f6c053e5bbcc/effekt/shared/src/main/scala/effekt/core/Tree.scala#L557) here?